WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift

```diff
@@ -111,9 +111,133 @@ final class OrderDetailsDataSource: NSObject {
         return shippingLines.first?.methodTitle ?? String()
     }
 
-    /// All the items inside an order
-    private var items: [OrderItem] {
-        return order.items
+    /// Yosemite.OrderItem
+    /// The original list of order items a user purchased
+    ///
+    private(set) var items: [OrderItem]
+
+    /// Combine refunded order items to show refunded products
+    ///
+    var refundedProducts: [RefundedProduct] {
+        /// OrderItemRefund.orderItemID isn't useful for finding duplicates
+        /// because multiple refunds cause orderItemIDs to be unique.
+        /// Instead, we need to find duplicate *Products*.
+        var items = [OrderItemRefund]()
+        for refund in refunds {
+            items.append(contentsOf: refund.items)
+        }
+
+        let currency = CurrencyFormatter()
+        // Creates an array of dictionaries, with the hash value as the key.
+        // Example: [hashValue: [item, item], hashvalue: [item]]
+        // Since dictionary keys are unique, this eliminates the duplicate `OrderItemRefund`s.
+        let grouped = Dictionary(grouping: items) { (item) in
+            return item.hashValue
+        }
+
+        return grouped.compactMap { (_, items) in
+            // Here we iterate over each group's items
+
+            // All items should be equal except for quantity and price, so we pick the first
+            guard let item = items.first else {
+                // This should never happen, but let's be safe
+                return nil
+            }
+
+            // Sum the quantities
+            let totalQuantity = items.sum(\.quantity)
+            // Sum the refunded product amount
+            let total = items
+                .compactMap({ currency.convertToDecimal(from: $0.total) })
+                .reduce(NSDecimalNumber(value: 0), { $0.adding($1) })
+
+            return RefundedProduct(
+                productID: item.productID,
+                variationID: item.variationID,
+                name: item.name,
+                price: item.price,
+                quantity: totalQuantity,
+                sku: item.sku,
+                total: total
+            )
+        }
+    }
+
+    /// Calculate the new order item quantities and totals after refunded products have altered the fields
+    ///
+    var aggregateOrderItems: [AggregateOrderItem] {
+        if refundedProducts.count == 0 {
+            fatalError("Error: attemtpted to calculate aggregate order item data with no refunded products.")
+        }
+
+        let currency = CurrencyFormatter()
+        // Convert the order items into a mutable type
+        var convertedItems = [AggregateOrderItem]()
```

Comment:
nit: instead of the for loop, can also use `map` so that it's easier to read and the array can be immutable:

```
let convertedItems = items.map { AggregateOrderItem(...) }
```

Reply:
It's still a `var` for convertedItems because map is appending the new types to the array. I like using `map`, though 💯